docarray/doc_index/backends/elastic_doc_index.py

```diff
@@ -0,0 +1,420 @@
+                    'dims': 128,
+                    'similarity': 'cosine',  # 'l2_norm', 'dot_product', 'cosine'
+                    'm': 16,
+                    'ef_construction': 100,
+                },
+                docarray.typing.ID: {'type': 'keyword'},
+                # `None` is not a Type, but we allow it here anyway
+                None: {},  # type: ignore
+                # TODO: add support for other types
+            }
+        )
+
+    ###############################################
+    # Implementation of abstract methods          #
+    ###############################################
+
+    def python_type_to_db_type(self, python_type: Type) -> Any:
+        """Map python type to database type."""
+        for allowed_type in ELASTIC_PY_VEC_TYPES:
+            if issubclass(python_type, allowed_type):
+                return np.ndarray
+
+        if python_type == docarray.typing.ID:
+            return docarray.typing.ID
+
+        raise ValueError(f'Unsupported column type for {type(self)}: {python_type}')
+
+    def _index(self, column_to_data: Dict[str, Generator[Any, None, None]]):
+
+        data = self._transpose_col_value_dict(column_to_data)
+        requests = []
+
+        for row in data:
+            request = {
+                '_index': self._db_config.index_name,
+                '_id': row['id'],
+            }
+            # TODO change here when more types are supported
+            for col_name, col in self._column_infos.items():
+                if not col.config:
+                    continue
+                if col.db_type == np.ndarray and np.all(row[col_name] == 0):
+                    row[col_name] = row[col_name] + 1.0e-9
+                request[col_name] = row[col_name]
+            requests.append(request)
+
+        _, warning_info = self._send_requests(requests)
+        for info in warning_info:
+            warnings.warn(str(info))
+
+        self._refresh(
+            self._db_config.index_name
+        )  # TODO add runtime config for efficient refresh
+
+    def num_docs(self) -> int:
+        return self._client.count(index=self._db_config.index_name)['count']
+
+    def _del_items(self, doc_ids: Sequence[str]):
+        requests = []
+        for _id in doc_ids:
+            requests.append(
+                {'_op_type': 'delete', '_index': self._db_config.index_name, '_id': _id}
+            )
+
+        _, warning_info = self._send_requests(requests)
+
+        # raise warning if some ids are not found
+        if warning_info:
+            ids = [info['delete']['_id'] for info in warning_info]
+            warnings.warn(f'No document with id {ids} found')
+
+        self._refresh(self._db_config.index_name)
+
+    def _get_items(self, doc_ids: Sequence[str]) -> Sequence[TSchema]:
+        accumulated_docs = []
+        accumulated_docs_id_not_found = []
+
+        for pos in range(0, len(doc_ids), MAX_ES_RETURNED_DOCS):
+
+            es_rows = self._client.mget(
```

Comment:
Users should ideally use the [Scroll API](https://www.elastic.co/guide/en/elasticsearch/reference/8.6/paginate-search-results.html#scroll-search-results) for fetching large results. We should throw an error if someone is using the `_get_items` for fetching all items or items more than some X% of the data size. I believe the `_get_items` must be a performant fetch using the current logic.

Reply:
I removed the `MAX_ES_RETURNED_DOCS` here because I notice there already is a `max_result_window` in ElasticSearch. I  don't know when should we throw the error, like what if there are only 20 documents and the user wants them all.